olp-cpp-sdk-dataservice-read/src/StreamLayerClientImpl.cpp

```diff
@@ -227,6 +230,62 @@ StreamLayerClientImpl::Unsubscribe() {
       std::move(cancel_token), std::move(promise));
 }
 
+client::CancellationToken StreamLayerClientImpl::GetData(
+    const model::Message& message, DataResponseCallback callback) {
+  auto get_data_task =
+      [=](client::CancellationContext context) -> DataResponse {
+    if (message.GetData()) {
+      OLP_SDK_LOG_INFO(kLogTag, "GetData: message contains the Data embedded");
+      return message.GetData();
+    }
+
+    const auto& data_handle = message.GetMetaData().GetDataHandle();
+
+    if (!data_handle) {
+      OLP_SDK_LOG_WARNING(kLogTag,
+                          "GetData: message does not contain data handle");
+      return client::ApiError(client::ErrorCode::InvalidArgument,
+                              "Data handle is missing");
+    }
+
+    OLP_SDK_LOG_TRACE_F(kLogTag, "GetData: started, data_handle=%s",
+                        data_handle.value().c_str());
+
+    auto blob_api = ApiClientLookup::LookupApi(
+        catalog_, context, kBlobService, kBlobVersion,
+        FetchOptions::OnlineIfNotFound, settings_);
+
+    if (!blob_api.IsSuccessful()) {
+      return blob_api.GetError();
+    }
+
+    const auto blob_response =
+        BlobApi::GetBlob(blob_api.GetResult(), layer_id_, data_handle.value(),
+                         boost::none, boost::none, context);
+
+    OLP_SDK_LOG_TRACE_F(kLogTag,
+                        "GetData: done, blob_response is successful: %s",
+                        blob_response.IsSuccessful() ? "true" : "false");
+
+    return blob_response;
+  };
+
+  return AddTask(settings_.task_scheduler, pending_requests_,
+                 std::move(get_data_task), std::move(callback));
+}
+
+client::CancellableFuture<DataResponse> StreamLayerClientImpl::GetData(
+    const model::Message& message) {
+  auto promise = std::make_shared<std::promise<DataResponse>>();
+  auto cancel_token =
+      GetData(std::move(message), [promise](DataResponse response) {
```

Comment:
```suggestion
      GetData(message, [promise](DataResponse response) {
```
The `move` here doesn't move, does it? Correct me if I'm wrong, but `message` is a const ref `lvalue` and cannot be cast to `const model::Message&&`, hence cannot be moved and needs to be copied.

Reply:
You are right, thanks.